cli/models.js

```diff
@@ -168,56 +168,99 @@ async function prepareModelRecord (db, model) {
   const id = model.id;
   const hasTimesteps = model.timesteps && model.timesteps.length;
 
-  // Filters to keep.
   let filters = [];
-  for (let filter of model.filters) {
-    filter.timestep = hasTimesteps ? filter.timestep === true : false;
-    if (filter.type === 'range') {
-      const res = await db.raw(
-        `
+
+  // If filters are defined, perform validation
+  if (Array.isArray(model.filters) && model.filters.length > 0) {
+    // Range filters need min/max values, which are calculated from
+    // scenarios output. The block below generate a query to fetch these values.
+    const rangeFilters = model.filters.filter(f => f.type === 'range');
+    if (rangeFilters.length > 0) {
+      let filterMinMaxSelectStrings = [];
+      let filterKeys = {};
+
+      for (let filter of rangeFilters) {
+        // Filter is time-stepped?
+        filter.timestep = hasTimesteps ? filter.timestep === true : false;
+
+        let filterCastStrings = [];
+        if (filter.timestep) {
+          // When range filter is time-stepped, min/max values should be
+          // calculated using all timesteps available.
+          for (const timestep of model.timesteps) {
+            const key = filter.key + timestep;
+            filterCastStrings.push(`CAST("filterValues" ->> :${key} AS FLOAT)`);
+            filterKeys[key] = key;
+          }
+        } else {
+          // When not time-stepped, use a filter key unmodified
+          filterCastStrings.push(
+            `CAST("filterValues" ->> :${filter.key} AS FLOAT)`
+          );
+          filterKeys[filter.key] = filter.key;
+        }
+
+        // Keep filter list
+        filterMinMaxSelectStrings.push(`
+          MIN(LEAST(${filterCastStrings.join(',')})) as "${filter.key}min",
```

Comment:
If I get this correctly you're getting the min and max across all timesteps.

for example, say we have as timesteps 2010, 2020, 2030. your min will be the minimum of 2010 and the max will be the maximum of 2030. Since these filters are timestepped, shouldn't there be a min and a max for every timestep?
It could happen that the minimum value of 2020 is lower that the minimum of 2010, or not?

Reply:
However I realize that having multiple min/max ranges for a filter would require quite some changes.

Maybe we need to calculate the min/max for every timestep and then get the absolute min/max from those?

Reply:
That is exactly what the code is supposed to be doing now. It gets the min/max values between time steps for every record with functions LEAST()/GREATEST(), then the min/max overall using aggregation funtions MIN()/MAX(). I agree with you it would require a great deal of changes to keep min/max for every timestep, thats why I've implemented the approach of getting the min/max overall.

Reply:
👍